feat(lsp): display codelens as virtual lines (instead of virtual text)

mikejmcguirk opened the pull request:
Problem: The codelens module displays lenses as virtual text. Because extmarks are set in the local display_line_lenses function, this cannot be changed directly. (Per igorlfs, this would close https://github.com/neovim/neovim/issues/33923)

Proposed solution: Create a display option in the codelens module that allows configuring which extmarks are set. Add code to support virtual lines.

<img width="590" height="198" alt="image" src="https://github.com/user-attachments/assets/3f9b699c-ae01-4dc7-bc49-fc46f15ac97d" />

Questions:
- ~~Is this the best design pattern to achieve this goal? I based it on vim.lsp.log.get_level() and set_level(). I wanted to avoid exposing more of the module's internals in the name of achieving a simple end result~~
- ~~Is CodeLensDisplay the best naming choice for the opts table?~~
- ~~Is the validation/annotation for hl_mode done in the best way? It's a direct copy of what's in the vim.api.keyset.set_extmark annotation, but would also have to be manually edited if the API changed~~

mikejmcguirk commented:
UPDATE:

- In line with vim.lsp.codelens.refresh.Opts, renamed the display opts to vim.lsp.codelens.display.Opts
- Added annotations and re-built docs
- Still unsure of how to handle the hl_mode option. An alternative could be to allow for passing in an entire extmark opts table, then letting display_line_lenses add the options it needs to using a table extend. The upside - very extensible and hackable. The downsides - Might be a lot to throw at an inexperienced user, and the errors on a malformed keyset would be brutal

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -1938,6 +1938,20 @@ Client:supports_method({method}, {bufnr})           *Client:supports_method()*
+    Display options for lenses
+
+    Fields: ~
+      • {extmark_params}  (`vim.api.keyset.set_extmark`) See
```
This makes me thing a callback might fit better. If we are just forwarding parameters to `nvim_buf_set_extmark` that's a signal that maybe the user can define a function that calls `nvim_buf_set_extmark` or whatever.

mikejmcguirk replied:
This update has been made

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -1997,6 +2017,14 @@ save({lenses}, {bufnr}, {client_id})                 *vim.lsp.codelens.save()*
       • {client_id}  (`integer`)
 
+set_display({opts})                           *vim.lsp.codelens.set_display()*
+    Sets display options for lenses
```
this is yet another way of configuring stuff. look for existing patterns. we have `vim.foo.config` for various things, that's one way. the other common mechanism is `enable(..., {opts})`.

mikejmcguirk replied:
Could use config(), and have a nil opts table return return the current config the way vim.diagnostic does. This would also resolve the comment on get_display() below

mikejmcguirk replied:
This update has been made

mikejmcguirk commented on runtime/doc/lsp.txt:
```diff
@@ -1964,6 +1978,12 @@ get({bufnr})                                          *vim.lsp.codelens.get()*
         (`lsp.CodeLens[]`)
 
+get_display()                                 *vim.lsp.codelens.get_display()*
+    Gets current display options for lenses
```
This is now removed.

mikejmcguirk commented:
Wanted to keep testing simple rather than bloating things with a kitchen sink mentality. But definitely open to feedback there. 

If good to go, I can write up a news entry and remove the draft status.

mikejmcguirk commented:
Added news and removed draft status. Been building off this for a week or so without issue.

That said, if there are additional suggestions, happy to address.

mikejmcguirk commented on runtime/doc/lsp.txt:
```diff
@@ -1922,6 +1922,24 @@ Client:supports_method({method}, {bufnr})           *Client:supports_method()*
 ==============================================================================
 Lua module: vim.lsp.codelens                                    *lsp-codelens*
 
+*vim.lsp.codelens.Config*
```
I can understand that, given where the overall state of those discussions is.

What about passing an opt and/or callback directly through codelens.refresh() and codelens.display()? This is in line with how a lot of other LSP functions handle customizable behavior, and also helps with the per-config question because different callbacks can be mapped based on which buf is being attached to (allowing leverage of future buf local config options).

mikejmcguirk replied:
Should I make any movement toward allowing the on_codelens client-to-server response handler to be deprecated? Or should I assume that's the interface going forward? 

It calls the display function and is called by the refresh function, so a callback based approach to setting display has to run through it.

mikejmcguirk replied:
> Does that need to be decided now, i.e. does it change the shape of this PR? Would prefer to minimize the diff of this PR so we can focus on step 1 before pondering step 2. Feel free to create an issue to track that idea and sketch it out.

I agree that would be a lot for this PR. I'll make an update focusing on the callback change.

Once that's done, I'd be happy to keep after it. The tracking issue is a good idea.

mikejmcguirk committed:
feat(lsp): display codelenses as virtual lines

Problem: Code lenses currently dispay as virtual text on the same line
and after the relevant item. While the spec does not say how lenses
should be rendered, above the line is most typical. For longer lines,
lenses rendered as virtual text can run off the side of the screen.

Solution: Display lenses as virtual lines above the text.

Closes https://github.com/neovim/neovim/issues/33923

Credit: Ofseed

For the original implementation of virutal line display in the
re-designed Codelens module.

Co-authored-by: Yi Ming <ofseed@foxmail.com>

mikejmcguirk commented:
I have re-scoped this PR to only address the question of displaying virtual lines by default. As noted in the updated commit message, while the spec does not say this is a requirement, it is the most typical method, and reduces issues with lenses running off the side of the screen.

In light of discussions I've seen in a couple different topics on how to implement customization for LSP modules, it seems like a question that should be given a broader treatment.

Since news.txt already calls out the Codelens module as "re-implemented", I didn't update it further.

@ofseed:
- The tests, including the restored check for clearing extmarks after the end of the buffer, were extremely helpful
- I changed the indentation scheme to snap to the text rather than the LSP's col value. Feels more intuitive
- I did remove  the zb command on row 0. I found that, if I opened a file that spawns a lens on the first line, and immediately input "Ofoo", the cursor position moves to the left (giving me the output "oof"). Maybe that's because of something in my config, and you could add a mode check in the on_win function, but I'm concerned that won't be the last edge case that comes up

As before, I'm open to suggestions and feedback on this.

ofseed commented:
It seems you just copied the content from https://github.com/ofseed/neovim/commit/fcc39fe2be83c4a23c2d9b19ae3ec15703af1127 and then made changes from there? In this case, I think it would be better for me to create a PR myself.

> * I changed the indentation scheme to snap to the text rather than the LSP's col value. Feels more intuitive

The range returned by the LSP server will definitely be smarter than simple indentation. If a line has a CodeLens, we can tell it is indented from the column value returned by the LSP server; in other words, it will always be indented under this approach. The only issue might be if you feel it is too far to the right, such as the "implementation" in the tests, but I don't think such subjective feelings should be included in the implementation since this is intended for everyone. Since the LSP specification provides the `range` field, there will certainly be servers that use it to adjust the display position of their CodeLens. In other words, the fact that "implementation" aligns with the struct name instead of appearing at column 0 is a result of the server's own choice. We should not simply ignore fields defined in the LSP.

> I did remove the zb command on row 0. I found that, if I opened a file that spawns a lens on the first line, and immediately input "Ofoo", the cursor position moves to the left (giving me the output "oof"). Maybe that's because of something in my config, and you could add a mode check in the on_win function, but I'm concerned that won't be the last edge case that comes up

This indeed cannot be reproduced in my configuration.

mikejmcguirk commented:
> It seems you just copied the content from [ofseed@fcc39fe](https://github.com/ofseed/neovim/commit/fcc39fe2be83c4a23c2d9b19ae3ec15703af1127) and then made changes from there? In this case, I think it would be better for me to create a PR myself.
>

If there's a way for the maintainers to add you as an author, or you want to re-open it yourself, I wouldn't have a problem with that.

> > * I changed the indentation scheme to snap to the text rather than the LSP's col value. Feels more intuitive
> 
> The range returned by the LSP server will definitely be smarter than simple indentation. If a line has a CodeLens, we can tell it is indented from the column value returned by the LSP server; in other words, it will always be indented under this approach. The only issue might be if you feel it is too far to the right, such as the "implementation" in the tests, but I don't think such subjective feelings should be included in the implementation since this is intended for everyone. Since the LSP specification provides the `range` field, there will certainly be servers that use it to adjust the display position of their CodeLens. In other words, the fact that "implementation" aligns with the struct name instead of appearing at column 0 is a result of the server's own choice. We should not simply ignore fields defined in the LSP.

At least for me, when I tried it out and saw the indentations over on the side of the text, I was confused as to why at first until I realized they were aligned with the relevant document symbols. And at least for me it makes the left edge of the text feel ragged. Also, while we shouldn't be emulating VSCode per se, they do snap to indent, and that does create a broader baseline expectation.

On the other hand, you're right that not using the LSP return data feels a bit arbitrary.

I'm flexible either way.

mikejmcguirk commented:
> This looks much simpler, thanks for followup. @ofseed did you have changes in mind that require a separate PR? If so feel free to open one and we can continue there. Or does this LGTY ?
> 
> Is there anything here that requires a doc update, e.g. do we need to mention how to enable this (code example)? And a news.txt item?

I added a line calling this out in news.txt.

No other doc changes are necessary.

justinmk commented:
Not an option. It should be possible to have full control via hooks, or custom handlers. Don't know the specifics in this case, but https://github.com/neovim/neovim/pull/37626 may have hints.
